### PR TITLE
Add YAML module

### PR DESCRIPTION
**Pull request by lsagetlethias**
This a a port from nodeca/js-yaml.

Typing internally is quite messy but the endpoints themselves (`parse` and `stringify`) are ready to use.

### CODE REVIEW

**Comment by zekth** [comment]
Could you move it to encoding/yaml? Like toml and csv?

**Comment by bios21** [comment]
Yes will do

**Comment by zekth** [comment]
Also please add references to https://github.com/nodeca/js-yaml in headers of ported files like here : https://github.com/denoland/deno_std/blob/master/encoding/csv.ts#L1

**Comment by bios21** [comment]
@zekth this PR is now RFR, tell me if squash is needed.

File: encoding/yaml/parse_test.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
```
Comment (zekth) [inline encoding/yaml/parse_test.ts]: Those tests files are not covering all the features of the lib.
Could you port those from here: https://github.com/nodeca/js-yaml/tree/master/test ?

**Comment by zekth** [comment]
Some comments. Also summoning @bartlomieju for review.

File: encoding/yaml/Mark.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { repeat } from "./utils.ts";
```
Comment (bartlomieju) [inline encoding/yaml/Mark.ts]: There are some files with capitalized/camelCase style, can you change it to snake_case to keep it identical with rest of repo?

File: encoding/yaml/State.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { SchemaDefinition } from "./Schema.ts";
+import { DEFAULT_FULL_SCHEMA } from "./schema/mod.ts";
+
+export abstract class State {
+  constructor(public schema: SchemaDefinition = DEFAULT_FULL_SCHEMA) {}
+}
```
Comment (bartlomieju) [inline encoding/yaml/State.ts]: I like that these objects are so well organized but... splitting them into 4 files means there are 4 more deps to download. I'd consider putting them into single file.

File: encoding/yaml/Type.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  resolve?: (data: Any) => boolean;
+  construct?: (data: string) => Any;
+  instanceOf?: Any;
+  predicate?: (data: object) => boolean;
+  represent?: RepresentFn | ArrayObject<RepresentFn>;
+  defaultStyle?: StyleVariant;
+  styleAliases?: ArrayObject;
+}
+
+function checkTagFormat(tag: string): string {
```
Comment (bartlomieju) [inline encoding/yaml/Type.ts]: seems superfluous

File: encoding/yaml/error/YAMLError.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    protected mark: Mark | string = ""
+  ) {
+    super(ErrorKind.Other, `${message} ${mark}`);
+    this.name = this.constructor.name;
+  }
+
+  public toString(_compact: boolean): string {
+    return `${this.name}: ${this.message} ${this.mark}`;
+  }
+}
```
Comment (bartlomieju) [inline encoding/yaml/error/YAMLError.ts]: Since there's only one error type, it seems superfluous to create `error` directory. How about `yaml/error.ts`?

File: encoding/yaml/schema/core.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { Schema } from "../Schema.ts";
+import { json } from "./json.ts";
+
+// Standard YAML's Core schema.
+// http://www.yaml.org/spec/1.2/spec.html#id2804923
+//
+// NOTE: JS-YAML does not support schema-specific tag resolution restrictions.
+// So, Core schema has no distinctions from JSON schema is JS-YAML.
```
Comment (bartlomieju) [inline encoding/yaml/schema/core.ts]: Is this comment relevant for this port?

File: encoding/yaml/schema/default_full.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+// Also this schema is used as default base schema at `Schema.create` function.
+
+export const defaultFull = (Schema.SCHEMA_DEFAULT = new Schema({
+  include: [defaultSafe]
+  /* TODO: JS */
+  //   explicit: [
+  //     require('../type/js/undefined'),
+  //     require('../type/js/regexp'),
+  //     require('../type/js/function')
+  //   ]
```
Comment (bartlomieju) [inline encoding/yaml/schema/default_full.ts]: Can you trim down those types for now?

File: encoding/yaml/schema/json.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { Schema } from "../Schema.ts";
+import { bool, float, int, nil } from "../type/mod.ts";
+import { failsafe } from "./failsafe.ts";
+
+// Standard YAML's JSON schema.
+// http://www.yaml.org/spec/1.2/spec.html#id2803231
+//
+// NOTE: JS-YAML does not support schema-specific tag resolution restrictions.
```
Comment (bartlomieju) [inline encoding/yaml/schema/json.ts]: s/JS-YAML/This implementation

**Comment by bios21** [comment]
Thank you for your reviews. I'm on it ASAP. :ok_hand:

File: encoding/yaml/devDeps.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+import * as asserts from "../testing/asserts.ts";
```
Comment (bartlomieju) [inline encoding/yaml/devDeps.ts]: Is this file needed?

File: README.md [hunk 10]
```diff
@@ -35,6 +35,7 @@ Here are the dedicated documentations of modules:
 - [strings](strings/README.md)
 - [testing](testing/README.md)
 - [ws](ws/README.md)
+- [yaml](yaml/README.md)
```
[the remark this answers is not in the document]
Comment (bios21) [reply README.md]: This was on a previous commit and no longer exists.

File: tsconfig.json [hunk 11]
```diff
@@ -1,5 +1,5 @@
 {
-  "extends": "tsconfig.test.json",
+  "extends": "./tsconfig.test.json",
```
[the remark this answers is not in the document]
Comment (bios21) [reply tsconfig.json]: Was a need when I setup my dev environment with yarn / node and executing `eslint` without npx.
Rollbacked.

File: encoding/yaml/State.ts (hunk 3 above)
Comment (bios21) [reply encoding/yaml/State.ts]: Do you mean all 4 kinds of schema to be grouped as one file ?

File: encoding/yaml/error/YAMLError.ts (hunk 5 above)
Comment (bios21) [reply encoding/yaml/error/YAMLError.ts]: Was an anticipation made at the beginning of the port.

Done.

File: encoding/yaml/schema/core.ts (hunk 6 above)
Comment (bios21) [reply encoding/yaml/schema/core.ts]: Removed.

File: encoding/yaml/schema/default_full.ts (hunk 7 above)
Comment (bios21) [reply encoding/yaml/schema/default_full.ts]: Removed.

File: encoding/yaml/schema/json.ts (hunk 8 above)
Comment (bios21) [reply encoding/yaml/schema/json.ts]: Removed.

File: encoding/yaml/devDeps.ts (hunk 9 above)
Comment (bios21) [reply encoding/yaml/devDeps.ts]: Not anymore, I forgot to remove it after adding the module from 3rd party to std.

Removed.

File: encoding/yaml/Mark.ts (hunk 2 above)
Comment (bios21) [reply encoding/yaml/Mark.ts]: All files are now snake_cased.

File: encoding/yaml.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+// Ported from js-yaml:
+// https://github.com/nodeca/js-yaml/
```
[the remark this answers is not in the document]
Comment (bios21) [reply encoding/yaml.ts]: Done on all files.

File: encoding/yaml/parse_test.ts (hunk 1 above)
Comment (bios21) [reply encoding/yaml/parse_test.ts]: JS files copied. Will do the conversion in ts/deno asap.

Comment (bios21) [reply encoding/yaml/parse_test.ts]: Will also be included issues specific tests from js-yaml repository.

File: encoding/yaml/type/binary.ts [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    result.push((bits >> 8) & 0xff);
+    result.push(bits & 0xff);
+  } else if (tailbits === 18) {
+    result.push((bits >> 10) & 0xff);
+    result.push((bits >> 2) & 0xff);
+  } else if (tailbits === 12) {
+    result.push((bits >> 4) & 0xff);
+  }
+
+  return new Buffer(new Uint8Array(result));
```
Comment (GrosSacASac) [inline encoding/yaml/type/binary.ts]: does Deno have Buffer ?
why not just return new Uint8Array(result) ?
